scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala

```diff
@@ -0,0 +1,379 @@
+import com.twitter.scalding.serialization.JavaStreamEnrichments
+
+trait LowerPriorityImplicit {
+  implicit def primitiveOrderedBufferSupplier[T] = macro com.twitter.scalding.macros.impl.OrderedSerializationProviderImpl[T]
+}
+
+object ByteBufferArb {
+  implicit def arbitraryTestTypes: Arbitrary[ByteBuffer] = Arbitrary {
+    for {
+      aBinary <- Gen.alphaStr.map(s => ByteBuffer.wrap(s.getBytes("UTF-8")))
+    } yield aBinary
+  }
+}
+object TestCC {
+  import ByteBufferArb._
+  implicit def arbitraryTestCC: Arbitrary[TestCC] = Arbitrary {
+    for {
+      aInt <- arb[Int]
+      aLong <- arb[Long]
+      aDouble <- arb[Double]
+      anOption <- arb[Option[Int]]
+      anStrOption <- arb[Option[String]]
+      anOptionOfAListOfStrings <- arb[Option[List[String]]]
+      aBB <- arb[ByteBuffer]
+    } yield TestCC(aInt, aLong, anOption, aDouble, anStrOption, anOptionOfAListOfStrings, aBB)
+  }
+}
+case class TestCC(a: Int, b: Long, c: Option[Int], d: Double, e: Option[String], f: Option[List[String]], aBB: ByteBuffer)
+
+object MyData {
+  implicit def arbitraryTestCC: Arbitrary[MyData] = Arbitrary {
+    for {
+      aInt <- arb[Int]
+      anOption <- arb[Option[Long]]
+    } yield new MyData(aInt, anOption)
+  }
+}
+
+class MyData(override val _1: Int, override val _2: Option[Long]) extends Product2[Int, Option[Long]] {
+  override def canEqual(that: Any): Boolean = that match {
+    case o: MyData => this._1 == o._1 && this._2 == o._2
+    case _ => false
+  }
+}
+
+object Container {
+  implicit def arbitraryInnerCaseClass: Arbitrary[InnerCaseClass] = Arbitrary {
+    for {
+      anOption <- arb[Set[Double]]
+    } yield InnerCaseClass(anOption)
+  }
+
+  type SetAlias = Set[Double]
+  case class InnerCaseClass(e: SetAlias)
+}
+class MacroOrderingProperties extends FunSuite with PropertyChecks with ShouldMatchers with LowerPriorityImplicit {
+  type SetAlias = Set[Double]
+
+  import ByteBufferArb._
+  import Container.arbitraryInnerCaseClass
+
+  def serialize[T](t: T)(implicit orderedBuffer: OrderedSerialization[T]): InputStream =
+    serializeSeq(List(t))
+
+  def serializeSeq[T](t: Seq[T])(implicit orderedBuffer: OrderedSerialization[T]): InputStream = {
+    import JavaStreamEnrichments._
+
+    val baos = new ByteArrayOutputStream
+    t.foreach({ e =>
+      orderedBuffer.write(baos, e)
+    })
+    baos.toInputStream
+  }
+
+  def rt[T](t: T)(implicit orderedBuffer: OrderedSerialization[T]) = {
+    val buf = serialize[T](t)
+    orderedBuffer.read(buf).get
+  }
+
+  def rawCompare[T](a: T, b: T)(implicit obuf: OrderedSerialization[T]): Int = {
+    obuf.compareBinary(serialize(a), serialize(b)).unsafeToInt
+  }
+
+  def oBufCompare[T](a: T, b: T)(implicit obuf: OrderedSerialization[T]): Int = {
+    obuf.compare(a, b)
+  }
+
+  def checkManyExplicit[T](i: List[T])(implicit obuf: OrderedSerialization[T]) = {
+    val serializedA = serializeSeq(i)
+    val serializedB = serializeSeq(i)
```

Comment:
this is weaker than taking `List[(T, T)]` unzipping them and comparing in binary being the same as in memory. In yours you are always exercising the equality check, which always has to read all the bytes no matter what.  I'd like to see this strengthened.